stdlib/public/core/NFC.swift

```diff
@@ -0,0 +1,223 @@
+      return nil
+    }
+
+    return Unicode.Scalar(_value: composition)
+  }
+
+  @inline(never)
+  internal func composeHangul(
+    _ x: Unicode.Scalar,
+    and y: Unicode.Scalar
+  ) -> Unicode.Scalar? {
+    // L = Hangul leading consonants
+    let L: (base: UInt32, count: UInt32) = (base: 0x1100, count: 19)
+    // V = Hangul vowels
+    let V: (base: UInt32, count: UInt32) = (base: 0x1161, count: 21)
+    // T = Hangul tail consonants
+    let T: (base: UInt32, count: UInt32) = (base: 0x11A7, count: 28)
+    // N = Number of precomposed Hangul syllables that start with the same
+    //     leading consonant. (There is no base for N).
+    let N: (base: UInt32, count: UInt32) = (base: 0x0, count: 588)
+    // S = Hangul precomposed syllables
+    let S: (base: UInt32, count: UInt32) = (base: 0xAC00, count: 11172)
+
+    switch (x.value, y.value) {
+    // Check for Hangul (L, V) -> LV compositions.
+    case (L.base ..< L.base &+ L.count, V.base ..< V.base &+ V.count):
+      let lIdx = x.value &- L.base
+      let vIdx = y.value &- V.base
+      let lvIdx = lIdx &* N.count &+ vIdx &* T.count
+      let s = S.base &+ lvIdx
+      return Unicode.Scalar(_value: s)
+
+    // Check for Hangul (LV, T) -> LVT compositions.
+    case (S.base ..< S.base &+ S.count, T.base &+ 1 ..< T.base &+ T.count):
+      if (x.value &- S.base) % T.count == 0 {
+        return Unicode.Scalar(_value: x.value &+ y.value &- T.base)
+      } else {
+        fallthrough
+      }
+
+    default:
+      return nil
+    }
+  }
+
+  internal mutating func next() -> Unicode.Scalar? {
+    // Empty out our buffer before attempting to compose anything with our new
+    // composee.
+    if let nextBuffered = buffer.next() {
+      return nextBuffered.scalar
+    }
+
+    while let current = iterator.next() {
+      guard let currentComposee = composee else {
+        // If we don't have a composee at this point, we're most likely looking
+        // at the start of a string. If our class is 0, then attempt to compose
+        // the following scalars with this one. Otherwise, it's a one off scalar
+        // that needs to be emitted.
+        if current.normData.ccc == 0 {
+          composee = current.scalar
+          continue
+        } else {
+          return current.scalar
+        }
+      }
+
+      // If we have any scalars in the buffer, it means those scalars couldn't
+      // compose with our composee to form a new scalar. However, scalars
+      // following them may still compose with our composee, so take the last
+      // scalar in the buffer and get its normalization data so that we can
+      // perform the check underneath this one about whether this current scalar
+      // is "blocked". We get the last scalar because the scalars we receive are
+      // already NFD, so the last scalar in the buffer will have the highest
+      // CCC value in this normalization segment.
+      guard let lastBufferedNormData = buffer.last?.normData else {
```

Comment:
It took me a few readings to figure it out, but we're really checking the prior scalar's CCC, right? So maybe:

```suggestion
guard let priorCCC = buffer.last?.normData.ccc else {
```

And update the comment appropriately.

Reply:
It's not necessarily the prior scalar's CCC, no. I can say something like `lastBufferedCCC` however and add the `.ccc` here.

Reply:
Isn't the prior CCC what's required by Unicode? When is it a different scalar?

"last" sounds like it's the last scalar in the current normalization segment, and thus would probably appear after `current`.